File: 208.implement-trie-prefix-tree.py
#
# @lc app=leetcode id=208 lang=python3
#
# [208] Implement Trie (Prefix Tree)
#

# @lc code=start
class TrieNode():
    def __init__(self,c = None,is_leaf=False):
        self.charactor = c
        self.children = {}
        self.is_leaf = False
class Trie:

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.root = TrieNode()

    def insert(self, word: str) -> None:
        """
        Inserts a word into the trie.
        """
        current = self.root
        found = True
        for index,c in enumerate(word):
            if current.children.get(c):
                current = current.children.get(c)
                continue
            else:
                found = False
                break
        if found: 
            current.is_leaf = True
        else:
            for c in word[index:]:
                node = TrieNode(c=c)
                current.children[c] = node
                current = current.children[c]
        current.is_leaf = True


    def search(self, word: str) -> bool:
        """
        Returns if the word is in the trie.
        """
        current = self.root
        for char in word:
            if current.children.get(char):
                current = current.children.get(char)
            else:
                return False
        return True if current and current.is_leaf else False

    def startsWith(self, prefix: str) -> bool:
        """
        Returns if there is any word in the trie that starts with the given prefix.
        """
        current = self.root
        for char in prefix:
            if current.children.get(char):
                current=current.children.get(char)
                continue
            else:
                return False
        return True 

# Your Trie object will be instantiated and called as such:
# obj = Trie()
# obj.insert(word)
# param_2 = obj.search(word)
# param_3 = obj.startsWith(prefix)
# @lc code=end

if __name__ == "__main__":
    root = Trie()
    root.insert("hello")
    print(root)

File: 208.implement-trie-prefix-tree.go
/*
 * @lc app=leetcode id=208 lang=golang
 *
 * [208] Implement Trie (Prefix Tree)
 */
package main

// some keys
// 1. how to identify end of an word
// 2. what data structure to specify the word
// 3. deal with search and startWith
// @lc code=start
type Trie struct {
	nodes map[string]*Trie
}

func Constructor() Trie {
	return Trie{
		nodes: map[string]*Trie{},
	}
}

func (this *Trie) Insert(word string) {
	cur := this
	for i := 0; i < len(word); i++ {
		if next, ok := cur.nodes[string(word[i])]; ok {
			cur = next
		} else {
			node := &Trie{
				//  to avoid nil map
				nodes: map[string]*Trie{},
			}
			cur.nodes[string(word[i])] = node
			cur = node
		}
	}
	cur.nodes["/"] = &Trie{}
}

func (this *Trie) Search(word string) bool {
	cur := this
	var ok bool
	for i := 0; i < len(word); i++ {
		if cur, ok = cur.nodes[string(word[i])]; !ok {
			return false
		}
	}
	_, ok = cur.nodes["/"]
	return ok
}

func (this *Trie) StartsWith(prefix string) bool {
	cur := this
	var ok bool
	for i := 0; i < len(prefix); i++ {
		if cur, ok = cur.nodes[string(prefix[i])]; !ok {
			return false
		}
	}
	return true
}

/**
 * Your Trie object will be instantiated and called as such:
 * obj := Constructor();
 * obj.Insert(word);
 * param_2 := obj.Search(word);
 * param_3 := obj.StartsWith(prefix);
 */
// @lc code=end
